sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnType.scala

```diff
@@ -34,7 +34,8 @@ import org.apache.spark.unsafe.types.UTF8String
  *
  * @tparam JvmType Underlying Java type to represent the elements.
  */
-private[sql] sealed abstract class ColumnType[JvmType] {
+private[sql]
+sealed abstract class ColumnType[@specialized(Boolean, Byte, Short, Int, Long) JvmType] {
```

Comment:
Haven't checking the JITed code, not sure the generated code will look like.

@rxin Will this work as expected (having specialized class for these primitive types)?